### PR TITLE
Update Database Connection to ODBC 3 to better support wide character strings

### LINKED ISSUE #768: Database Connection - Incorrect buffer size for wide string columns under some databases/drivers
I'm updating my ReStore database framework following the recent changes to DBConnection and am running into a few issues with wide (UTF16) string columns.

Under MSAccess and SQLite ODBC drivers the size of wide string columns is returned as their character size, not the storage size. This leads to a bounds error when recovering data from those columns. Example:
```
connection := DBConnection new.
connection dsn: 'AnAccessDSN'.
connection connect.

connection exec: 'create table utf16_test (a_string VARCHAR (8))'.
(connection exec: 'select * from utf16_test') describeCols: #(1).  
"#(a DBColAttr(1, a_string, SQL_WVARCHAR, 8))"

connection exec: 'insert into utf16_test(a_string) values(''12345678'')'.
(connection exec: 'select * from utf16_test') results next  
"Index 16 is out of bounds"
```
The issue doesn't present with the other databases/drivers I've tested (SQLServer Express, Postrgres, MySQL).

Looking at senders of `sqlColAttribute:columnNumber:...` the issue can partly be traced to the fieldIdentifier parameter of `SQL_COLUMN_LENGTH` which has the comment `"Note using ODBC 2.x definition of column length"`. I can't find the relevant documentation for this parameter, however the documentation for the ODBC 3 near-equivalent of `SQL_DESC_LENGTH` (1003) states: 

> A numeric value that is either the maximum or actual character length of a character string or binary data type. 

i.e. the character length, not the storage length. Using this parameter instead of `SQL_COLUMN_LENGTH` displays the same issue.

The ODBC 3 parameter `SQL_DESC_OCTET_LENGTH` (1013; I can't find an ODBC 2.0 version of this) would appear to be the correct one to use; the documentation states:

> The length, in bytes, of a character string or binary data type.

Indeed, changing to this parameter fixes the wide string issue for both Access and SQLite. Unfortunately this causes a different issue for Access DATETIME fields:
```
connection exec: 'create table datetime_test (a_timestamp DATETIME)'.
(connection exec: 'select * from datetime_test') describeCols: #(1). 
"With SQL_COLUMN_LENGTH   		#(a DBColAttr(1, a_timestamp, SQL_TIMESTAMP, 16))"
"With SQL_DESC_OCTET_LENGTH		#(a DBColAttr(1, a_timestamp, SQL_TIMESTAMP, 8))"
```

Recovering DATETIME fields with this value results in an undersized buffer and usually the image crashing out. 

The solution I'm using for now is to use `SQL_DESC_OCTET_LENGTH` when dealing with string or binary columns (as per the documentation) and `SQL_COLUMN_LENGTH` for all other column types - this works for Access and the other databases I've tried. 

I'd like to submit this fix however I'm a little uncomfortable mixing parameters from different ODBC versions. I've tried using `SQL_DESC_LENGTH` instead of `SQL_COLUMN_LENGTH` for non-string columns however this causes Access to return 19 (the column precision) as the size of DATETIME columns - which works since it delivers a big enough buffer, but is technically incorrect.

I also note that the documentation for both `SQL_DESC_LENGTH` and `SQL_DESC_OCTET_LENGTH` only refer to their use with string/binary columns, though they do also work (Access oddities aside) with other column types. I'd be interested to see the documentation for `SQL_COLUMN_LENGTH` if anyone can locate it.

### PR DESCRIPTION
**Pull request by rko281**
- Set `SQL_OV_ODBC3` in `DBConnection class>>setDefaultEnvAttrs`
 - Updated `ODBCConstants`(etc.) with ODBC 3 declarations
 - Use `SQL_DESC_OCTET_LENGTH` to determine variable-sized (string or binary) field widths.
 - Fixed-size fields now determine their size based on their type (`DBColAttr>>typeLength`).

Tested against the ReStore test suite with SQLite, MySQL, Access and SQL Server with a mix of ANSI and wide character type declarations and/or driver settings. All run as expected. Have also updated the small Database Connection test package to reflect Access using wide character fields, and run the unicode tests from #752

Closes #768

### CODE REVIEW

File: Core/Object Arts/Dolphin/Database/DBColAttr.cls [hunk 1]
```diff
@@ -198,6 +209,31 @@ type: anInteger
...
 
+typeLength
+
+	"Private - Return the length (size in bytes) of the receiver based on its type."
+
+	| cType |
+
+	cType := self cType.
+
+	(cType = SQL_C_BIT or: [cType = SQL_C_TINYINT]) ifTrue: [^1].
```
[the remark this answers is not in the document]
Comment (rko281) [reply Core/Object Arts/Dolphin/Database/DBColAttr.cls]: Hmmm... I initially coded this using the CTypesExtraLength as per your original suggestion. This had a minor issue with the printString showing zero size (easily fixed) but I also (I thought) hit an issue during the ReStore test runs with zero returned as the octet length for a DECIMAL column under (I think) Postgres. At this point I rewrote as the "switch" including the manual calculation of the decimal size. However as you point out below this code path is never actually taken so I was clearly mistaken in what I thought I saw/fixed.

File: Core/Object Arts/Dolphin/Database/DBColAttr.cls [hunk 2]
```diff
@@ -198,6 +209,31 @@ type: anInteger
...
+	cType = SQL_C_DOUBLE ifTrue: [^8].
+	(cType = SQL_C_TIMESTAMP or: [cType = SQL_C_GUID]) ifTrue: [^16].
+
+	"SQL_BIGINT
+	The number of bytes required to hold the character representation of this data if the character set is ANSI, and twice this number if the character set is UNICODE, because this data type is returned as a character string by default. The character representation consists of 20 characters: 19 for digits and a sign, if signed, or 20 digits, if unsigned. Therefore, the length is 20."
+	self type = SQL_BIGINT ifTrue: [^##(20 * 2 "allow for unicode")].
+
+	"SQL_DECIMAL, SQL_NUMERIC
+	The number of bytes required to hold the character representation of this data if the character set is ANSI, and twice this number if the character set is UNICODE. This is the maximum number of digits plus two, because the data is returned as a character string and characters are needed for the digits, a sign, and a decimal point. For example, the transfer length of a column defined as NUMERIC(10,3) is 12."
+	(self type = SQL_DECIMAL or: [self type = SQL_NUMERIC]) ifTrue: [^self length: (self precision + 2) * 2 "allow for unicode"].
```
[the remark this answers is not in the document]
Comment (rko281) [reply Core/Object Arts/Dolphin/Database/DBColAttr.cls]: Left over from an earlier version of the method which set the length rather than returning it. 

I'll tidy all this up and retest at some point.

File: Core/Object Arts/Dolphin/Database/DBConnectionTest.cls [hunk 3]
```diff
@@ -1,4 +1,4 @@
-"Filed out from Dolphin Smalltalk 7"!
+﻿"Filed out from Dolphin Smalltalk 7"!
```
[the remark this answers is not in the document]
Comment (rko281) [reply Core/Object Arts/Dolphin/Database/DBConnectionTest.cls]: Indeed - I eventually managed to track down a version which largely works with the tests (aside from a differently-configured set of views).

File: Core/Object Arts/Dolphin/Database/DBField.cls [hunk 4]
```diff
@@ -471,13 +471,16 @@ initialize
 				at: SQL_DATE + TypeOffset put: #getDate;
 				at: SQL_TIME + TypeOffset put: #getTime;
 				at: SQL_TIMESTAMP + TypeOffset put: #getDateAndTime;
+				at: SQL_TYPE_DATE + TypeOffset put: #getDate;
```
[the remark this answers is not in the document]
Comment (rko281) [reply Core/Object Arts/Dolphin/Database/DBField.cls]: No idea, though I now notice the C type constants have also changed in the same way. I'll tidy this up too.
